packages/components/src/segmented-control/README.md

```diff
@@ -0,0 +1,87 @@
+# SegementedControl
+
+<div class="callout callout-alert">
+This feature is still experimental. “Experimental” means this is an early implementation subject to drastic and breaking changes.
+</div>
+
+`SegementedControl` is a form component that lets users choose options represented in horizontal segments.
+
+## Usage
+
+```js
+import { __experimentalSegmentedControl as SegmentedControl } from '@wordpress/components';
+
+function Example() {
+	return (
+		<SegmentedControl
+			label="My Segmented Control"
+			value="vertical"
+			options={ [
+				{
+					label: 'Horizontal',
+					value: 'horizontal',
+				},
+				{
+					label: 'Vertical',
+					value: 'vertical',
+				},
+			] }
```

Comment:
Given @diegohaz's recommendation to move away from APIs like this (similar to the `controls` API in `DropdownMenu`) I think we should consider refactoring this component to include a `SegmentedControl.Option` that should be rendered as children, rather than this `options` array which is harder to use.

Instead it should look something like this:

```jsx
<SegmentedControl label="My Segmented Control" value="vertical">
  <SegmentedControl.Option value="horizontal" label="Horizontal" />
  <SegmentedControl.Option value="vertical" label="Vertical" />
</SegmentedControl>
```

Then the internal API would map over children using `Children.map` from `React` and pass the appropriate extra props:

```jsx
const mappedChildren = Children.map(children, (child, index) => {
  if (! isValidElement(child) || typeof child.type === 'string') {
    return child;
  }
  
  const showSeparator = getShowSeparator( radio, index );
  return cloneElement(child, { ...child.props, showSeparator, isBlock: ! isAdaptiveWidth, ...etc });
});
```

Or something like that at least.

Then `SegmentedControl.Option` would be something as simple as like just rendering a button:

```jsx
function Option(props, forwardedRef) {
  const buttonProps = useContextSystem(props, 'SegmentedControlOption');

  return <Button {...buttonProps} />
```

What do you think about this, Nik? Diego?

Reply:
Yeah! I agree that we should move away from the `controls` prop. But I wouldn't do the `Children.map` and `cloneElement` either as it prevents us from doing things like this:

```jsx
function Options() {
  return (
    <>
      <SegmentedControl.Option />
      <SegmentedControl.Option />
    </>
  );
}

<SegmentedControl>
  <Options />
</SegmentedControl>
```

We can use React Context to avoid that (but I guess `useContextSystem` also solves that, doesn't it?).

`SegmentedControl.Option` vs. `SegmentedControlOption` is another point of discussion. I'm not sure if we have more of the former or the latter in the library, but we should agree on a pattern here.

Reply:
> But I wouldn't do the Children.map and cloneElement either as it prevents us from doing things like this

Oh good call! I hadn't considered that.

Yeah I think we'd still have to somehow get the `index` prop into the option component otherwise it won't be able to call the `getShowSeparator` function 🤔 I'm not sure how best to do that using React context or useContextSystem as you can't really "target" a specific instance of a component for either of them 🤔

Reply:
We just have to pass the object returned by `useRadioState` down. This object keeps an array with the items in `radio.items`. It already takes into account lazily mounted elements and other cases where elements are re-ordered. So I guess we can get the index from there.

Reply:
Oh cool! We've been using `React.Context` for things like that mostly because it doesn't require adding a new prop, as `useContextSystem` only lets you modify the props that are passed to a component so you are necessarily dealing with their public API. And in this case because the `Option` component will always only be used for the `SegmentedControl` there is already a dependency there, so introducing a shared context isn't an issue.

Reply:
>SegmentedControl.Option vs. SegmentedControlOption is another point of discussion. I'm not sure if we have more of the former or the latter in the library, but we should agree on a pattern here.

`SegmentedControl.Option` seems much more fitting to me as it will only be used by SegmentedControl

Reply:
So, is this close to what you had in mind? https://github.com/WordPress/gutenberg/pull/31937/commits/54757743d954b332001a364b4945b66e11968783

Reply:
I'd appreciate a review when you find some time @diegohaz and @sarayourfriend about the approach. And if it works for you I can update the storybook and tests to land this.

Reply:
Yes! To me that looks great!

Reply:
Nice! I'll take a better look later today. I'd like to discuss more the naming convention (`SegmentedControl.Option` vs. `SegmentedControlOption`), especially in the light of #33391. It doesn't look like we have many modules using this dot notation if any. So we would need to change more things if we want to adopt this convention. For now, I would stick with `SegmentedControlOption`.